core-bundle/src/Controller/ContentElement/MarkdownController.php

```diff
@@ -0,0 +1,93 @@
+use Contao\Template;
+use League\CommonMark\Environment;
+use League\CommonMark\Extension\Autolink\AutolinkExtension;
+use League\CommonMark\Extension\ExternalLink\ExternalLinkExtension;
+use League\CommonMark\Extension\Strikethrough\StrikethroughExtension;
+use League\CommonMark\Extension\Table\TableExtension;
+use League\CommonMark\Extension\TaskList\TaskListExtension;
+use League\CommonMark\MarkdownConverter;
+use League\CommonMark\MarkdownConverterInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+
+class MarkdownController extends AbstractContentElementController
+{
+    /**
+     * @var MarkdownConverterInterface
+     */
+    private $converter;
+
+    public function __construct(MarkdownConverterInterface $converter)
+    {
+        $this->converter = $converter;
+    }
+
+    public static function createController(RequestStack $requestStack): self
+    {
+        $environment = Environment::createCommonMarkEnvironment();
+
+        // Support GitHub flavoured Markdown (using the individual extensions because we don't want the
+        // DisallowedRawHtmlExtension which is included by default)
+        $environment->addExtension(new AutolinkExtension());
+        $environment->addExtension(new StrikethroughExtension());
+        $environment->addExtension(new TableExtension());
+        $environment->addExtension(new TaskListExtension());
+
+        // Automatically mark external links as such if we have a request
+        if (null !== ($request = $requestStack->getCurrentRequest())) {
+            $environment->addExtension(new ExternalLinkExtension());
+            $environment->mergeConfig([
+                'external_link' => [
+                    'internal_hosts' => $request->getHost(),
+                    'open_in_new_window' => true,
+                    'html_class' => 'external-link',
+                    'noopener' => 'external',
+                    'noreferrer' => 'external',
+                ],
+            ]);
+        }
+
+        $converter = new MarkdownConverter($environment);
+
+        return new self($converter);
+    }
+
+    protected function getResponse(Template $template, ContentModel $model, Request $request): ?Response
+    {
+        $this->initializeContaoFramework();
+        $markdown = '';
+
+        if ('sourceFile' === $model->markdownSource) {
+            /** @var FilesModel|null $filesModel */
+            $filesModel = $this->get('contao.framework')->getAdapter(FilesModel::class)->findByPk($model->singleSRC);
+
+            if (null !== $filesModel) {
+                $markdown = (string) $filesModel->getContent();
+            }
+        } else {
+            $markdown = $model->code;
+        }
+
+        $html = $this->converter->convertToHtml($markdown);
+
+        return new Response(strip_tags($html, $this->get('contao.framework')->getAdapter(Config::class)->get('allowedTags')));
```

Comment:
Just thinking: If we wanted to open up the environment in the future (e.g. by using a factory or dispatching an event that allows adding your own extensions) it would probably be better to use a renderer to check for allowed tags. Then the output of other inline elements/blocks could very well safely use any tag but a user still could not inject them.

Reply:
Yeah, that's a new feature though. We haven't had that up until now and I don't need that so feel free, if you need it :)